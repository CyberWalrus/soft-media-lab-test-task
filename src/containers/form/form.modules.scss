.form {
  &__input {
    border: 0.1rem solid var(--border-color);
    border-radius: 2rem;
  }
}
