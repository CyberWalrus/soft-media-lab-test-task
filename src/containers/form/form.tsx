/* eslint-disable jsx-a11y/label-has-associated-control */
import React, { FunctionComponent } from 'react';
import { connect } from 'react-redux';
import { Field, reduxForm, InjectedFormProps, formValueSelector } from 'redux-form';
import Popup from '../../components/popup/popup';
import InfoBox from '../../components/info-box/info-box';
import Input from '../../components/input/input';
import typePay from '../../constants/type-pay';
import './form.modules.scss';

type TypeSalary = 'month' | 'mrot' | 'day' | 'house';
interface FormProps {
  typeSalary: TypeSalary;
  ndfl: boolean;
  salary: string;
}

interface FormData {
  typeSalary: TypeSalary;
  ndfl: boolean;
  salary: string;
}

const defaultValues: FormData = {
  typeSalary: 'month',
  ndfl: true,
  salary: '',
};

type Props = FormProps & InjectedFormProps<FormData, FormProps>;

const Form: FunctionComponent<Props> = ({ handleSubmit, ndfl, typeSalary, salary }: Props) => {
  return (
    <form className="form" onSubmit={handleSubmit}>
      <div className="form__title">Сумма</div>
      <div className="form__radio-box">
        <label className="form__radio-item">
          <Field name="typeSalary" component="input" type="radio" value="month" />
          Оклад за месяц
        </label>
        <div className="form__radio-item">
          <label className="form__radio">
            <Field name="typeSalary" component="input" type="radio" value="mrot" />
            МРОТ
          </label>
          <Popup>
            <div>
              <b>МРОТ</b>
              <span>- минимальный размер оплаты труда. Разный для разных регионов.</span>
            </div>
          </Popup>
        </div>
        <label className="form__radio-item">
          <Field name="typeSalary" component="input" type="radio" value="day" />
          Оплата за день
        </label>
        <label className="form__radio-item">
          <Field name="typeSalary" component="input" type="radio" value="house" />
          Оплата за час
        </label>
      </div>
      <div className="form__switch-box">
        <div className={`form__title${ndfl ? '' : ' form__title--active'}`}>Указать с НДФЛ</div>
        <label className="form__switch">
          <Field name="ndfl" id="employed" component="input" type="checkbox" />
          <span className="form__slider" />
        </label>
        <div className={`form__title${!ndfl ? '' : ' form__title--active'}`}>без НДФЛ</div>
      </div>
      {typeSalary !== 'mrot' && (
        <div className="form__input-box">
          <Field name="salary" component={Input} className="form__input" />
          <div>{typePay[typeSalary] || '₽'}</div>
        </div>
      )}
      {typeSalary === 'month' && <InfoBox salary={salary} ndfl={ndfl} />}
      <div />
    </form>
  );
};

export const FormConnected = reduxForm<FormData, FormProps>({
  initialValues: defaultValues,
  form: 'payForm',
})(Form);

const selector = formValueSelector('payForm');

export default connect((state) => {
  const ndfl = selector(state, 'ndfl');
  const salary = selector(state, 'salary');
  const typeSalary = selector(state, 'typeSalary');
  return {
    ndfl,
    salary,
    typeSalary,
  };
})(FormConnected);
