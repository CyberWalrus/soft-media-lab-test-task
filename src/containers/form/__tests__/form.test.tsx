import React from 'react';
import { Provider } from 'react-redux';
import { shallow } from 'enzyme';
import configureStore from 'redux-mock-store';
import { Store, AnyAction } from 'redux';
import { reducer as formReducer } from 'redux-form';
import { FormConnected } from '../form';

describe('Form component', () => {
  const initialState = { form: formReducer };
  const mockStore = configureStore();
  let store: Store<any, AnyAction>;

  beforeAll(() => {
    store = mockStore(initialState);
  });

  it('should renders correctly', () => {
    const wrapper = shallow(
      <Provider store={store}>
        <FormConnected typeSalary="mrot" ndfl salary={0} />
      </Provider>
    );
    expect(wrapper.isEmptyRender()).not.toEqual(true);

    expect(wrapper.html()).toMatchSnapshot();
  });
});
