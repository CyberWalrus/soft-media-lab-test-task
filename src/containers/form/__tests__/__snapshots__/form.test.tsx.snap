// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form component should renders correctly 1`] = `"<form class=\\"form\\"><div class=\\"form__title\\">Сумма</div><div class=\\"form__radio-box\\"><label class=\\"form__radio-item\\"><input type=\\"radio\\" name=\\"typeSalary\\" value=\\"month\\"/>Оклад за месяц</label><div class=\\"form__radio-item\\"><label class=\\"form__radio\\"><input type=\\"radio\\" name=\\"typeSalary\\" value=\\"mrot\\"/>МРОТ</label><div class=\\"popup\\"><button class=\\"popup__button\\" type=\\"button\\"><span class=\\"popup__button-text\\">i</span></button></div></div><label class=\\"form__radio-item\\"><input type=\\"radio\\" name=\\"typeSalary\\" value=\\"day\\"/>Оплата за день</label><label class=\\"form__radio-item\\"><input type=\\"radio\\" name=\\"typeSalary\\" value=\\"house\\"/>Оплата за час</label></div><div class=\\"form__switch-box\\"><div class=\\"form__title\\">Указать с НДФЛ</div><label class=\\"form__switch\\"><input type=\\"checkbox\\" name=\\"ndfl\\" value=\\"\\" id=\\"employed\\"/><span class=\\"form__slider\\"></span></label><div class=\\"form__title form__title--active\\">без НДФЛ</div></div><div></div></form>"`;
